### ISSUE TITLE
Probably deadloop in @nestjs/core/helpers/handler-metadata-storage or object-hash

### ISSUE THREAD

**Issue by beenotung** [nestjs/nest#3231]
## Current behavior

The nestjs application init process seems running into deadloop.
Sometime got `UnhandledPromiseRejectionWarning: Error: Unknown object type "promise"` from object-hash called from nestjs core, sometime just freeze there with high CPU usage.

## Input Code

The controller only has one api endpoint, the code are simplified for illustration. I've other project using the same controller code but it can run the nestjs application without issue.
```ts
import { Body, Controller, Post, Req, Res } from '@nestjs/common';
import { CallInput, isPromise, LogService, Result } from 'cqrs-exp';
import { Request, Response } from 'express-serve-static-core';
import { ok, rest_return } from 'nestlib';
import { Call } from '../domain/types';
import { CoreService } from './core.service';

@Controller('core')
export class CoreController {
  static instance: CoreController;
  logService: LogService;
  ready: Promise<void>;

  constructor(public coreService: CoreService) {
    CoreController.instance = this;
    // create on my own instead of using injection as workaround for cyclic dependency
    this.logService = new LogService('data/log');
    this.ready = this.restore();
  }

  async restore() {
    // ... do some state restore
    // ... listen to commands from primus.js
  }

  storeAndCall<C extends Call>(call: CallInput<C>): Result<C['Out']> {
    // log down the command
    this.logService.storeObjectSync(
      call,
      this.logService.nextKey() + '-' + call.CallType,
    );
    // run the command to get result
    return this.coreService.Call<C>(call);
  }

  @Post('Call')
  async Call<C extends Call>(
    @Req() req: Request,
    @Res() res: Response,
    @Body() call: CallInput<C>,
  ) {
    await this.ready;
    try {
      let out = this.storeAndCall<C>(call);
      if (isPromise(out)) {
        out = await out;
      }
      ok(res, out);
      return out;
    } catch (e) {
      return rest_return(res, Promise.reject(e));
    }
  }
}
```

## Console Output
```
[Nest] 298740   - 10/22/2019, 11:16:28 AM   [NestFactory] Starting Nest application...
[Nest] 298740   - 10/22/2019, 11:16:33 AM   [InstanceLoader] AppModule dependencies initialized +4311ms
[Nest] 298740   - 10/22/2019, 11:16:33 AM   [InstanceLoader] CoreModule dependencies initialized +0ms
[Nest] 298740   - 10/22/2019, 11:16:33 AM   [RoutesResolver] AppController {/}: +44ms
[Nest] 298740   - 10/22/2019, 11:16:33 AM   [RouterExplorer] Mapped {/, GET} route +4ms
[Nest] 298740   - 10/22/2019, 11:16:33 AM   [RoutesResolver] CoreController {/core}: +1ms
```

Below console output should appear as in other project but they doesn't show up, instead the node process keeps using high CPU usage without further output:
```
[Nest] 242217   - 10/22/2019, 10:55 AM   [RouterExplorer] Mapped {/Call, POST} route +7ms
[Nest] 242217   - 10/22/2019, 10:55 AM   [NestApplication] Nest application successfully started +3ms
```
(the time is different as it's copied from another project with the same API setup)

## Expected behavior

The nestjs application should be able to init without issue, then response to request from the client.

## Possible Solution

May need to look into `@nestjs/core/helpers/handler-metadata-storage` to further investigate.

## Environment

<pre><code>
Nest version: 6.10.5

 
For Tooling issues:
- Node version: v10.16.0  
- Platform: Linux 

Others:

I'm nvm 0.33.11 to manage npm
</code></pre>

**Comment by beenotung** [comment]
Attached the CPU profile collected from `node --inspect`
![1](https://user-images.githubusercontent.com/6007292/67258436-2b135c80-f4c3-11e9-9f1f-71a42c6d8e3e.png)

**Comment by beenotung** [comment]
the freezing bug can be solved if I modify object-hash as below but it doesn't seems like fixing the root cause.

node_modules/object-hash/index.js aroung line 223 (inside `_object: function(object) {...}`)
```ts
 if(objType !== 'object' && objType !== 'function') {
        if(this['_' + objType]) {
          this['_' + objType](object);
        } else if (options.ignoreUnknown) {
          return write('[' + objType + ']');
        } else {
          // modified the 'else' block
          if (objType === 'promise') {
            object.then(o=>{
              if(o === undefined){return} // this is the fix
              this._object(o) // this will cause "Error [ERR_CRYPTO_HASH_FINALIZED]: Digest already called" because object-hash doesn't support async usage
            })
          } else {
            throw new Error('Unknown object type "' + objType + '"'); // the original fallback action
          }
        }
      }
```

**Comment by beenotung** [comment]
It seems the cause is found:
In my case, the controller has a service,
the service has a class object instance,
that object load a large hashed array in the constructor.

If I make the array smaller (e.g. slice to 1000 elements), the nestjs application can start up normaly. However I need to load > 85,000 elements in that array.

It seems nestjs is applying recursive object-hash on the controller / service class properties. Why is it necessary?

**Comment by beenotung** [comment]
I'm quite sure it's nestjs's decorator doing the costly object-hash, if I refactor the class property to the global variable, it startup instantly.

I would consider this workaround for now, the class is a singletone anyway so should be fine.

**Comment by beenotung** [comment]
For other people facing the same problem, you can import this script from the first line in main.ts

I know it's not a good practice, but it's a temporary workaround before the bug in @nestjs/core is fixed

hotfix.ts
```ts
import { readFileSync, existsSync, writeFileSync } from 'fs';
import { join } from 'path';

function hotfix() {
  for (let filename of [
    join('node_modules', 'object-hash', 'index.js'),
    join('node_modules', '.pnpm', 'node_modules', 'object-hash', 'index.js'),
  ]) {
    if (existsSync(filename)) {
      let originalSource = '          throw new Error(\'Unknown object type "\' + objType + \'"\');';
      let fixedSource = '          if (objType === \'promise\') {';
      let text = readFileSync(filename).toString();
      let lines = text.split('\n');
      let line = lines[222];
      if (line === fixedSource) {
        return;
      }
      if (line !== originalSource) {
        continue;
      }
      line = `          if (objType === 'promise') {
            object.then(o=>{
              if(o===undefined){return}
              this._object(o);
            })
          } else {
            throw new Error('Unknown object type "' + objType + '"');
          }`;
      lines[222] = line;
      text = lines.join('\n');
      writeFileSync(filename, text);
      console.log('hotfix', filename, 'for bugs in @nestjs/core/helpers/handler-metadata-storage');
      return;
    }
  }
  console.error('object-hash module is not found');
  process.exit(1);
}

hotfix();
```

**Comment by kamilmysliwiec** [comment]
Fixed in 6.8.4. Let me know if you face any issues.
